wicketd/src/http_entrypoints.rs

```diff
@@ -896,21 +897,37 @@ async fn post_start_update(
     // 1. We haven't pulled its state in our inventory (most likely cause: the
     //    cubby is empty; less likely cause: the SP is misbehaving, which will
     //    make updating it very unlikely to work anyway)
-    // 2. We have pulled its state but our hardware manager says we can't update
-    //    it (most likely cause: the target is the sled we're currently running
-    //    on; less likely cause: our hardware manager failed to get our local
-    //    identifying information, and it refuses to update this target out of
-    //    an abundance of caution).
+    // 2. We have pulled its state but our hardware manager says we can't
+    //    update it (most likely cause: the target is the sled we're currently
+    //    running on; less likely cause: our hardware manager failed to get our
+    //    local identifying information, and it refuses to update this target
+    //    out of an abundance of caution).
     //
-    // First, get our most-recently-cached inventory view.
-    let inventory = match rqctx.mgs_handle.get_cached_inventory().await {
-        Ok(inventory) => inventory,
-        Err(ShutdownInProgress) => {
+    // First, get our most-recently-cached inventory view. (Only wait 80% of
+    // WICKETD_TIMEOUT for this -- if the inventory isn't available, then we
+    // should produce a useful error message rather than timing out on the
+    // client.)
+    let inventory = match tokio::time::timeout(
+        WICKETD_TIMEOUT.mul_f32(0.8),
+        rqctx.mgs_handle.get_cached_inventory(),
+    )
+    .await
+    {
+        Ok(Ok(inventory)) => inventory,
+        Ok(Err(ShutdownInProgress)) => {
             return Err(HttpError::for_unavail(
                 None,
                 "Server is shutting down".into(),
             ));
         }
+        Err(_) => {
+            // Use 400 Bad Request instead of 503 Service Unavailable so that
+            // the error message is propagated to the client.
+            return Err(HttpError::for_bad_request(
```

Comment:
We can (and probably should in this case, right?) return a 503 while still providing a client-visible error message; it's just more verbose: https://github.com/oxidecomputer/omicron/blob/ca9d90a79c2a7a9ffdca12a705d007d7f28ce61f/wicketd/src/http_entrypoints.rs#L389-L395

`external_message` specifically is returned to the client. In the [helper method](https://github.com/oxidecomputer/dropshot/blob/1e187aff50fc8559330c63e6693cc4deade22e28/dropshot/src/error.rs#L182-L185) it's set to "Service Unavailable", but the fields are intentionally pub to allow this, I believe.

Reply:
Ah interesting! Thanks for the pointer.